# Imports

In [ ]:
import kagglehub
import pandas as pd
import numpy as np
import os

# Loading Data

In [ ]:
#Downloading data from Kaggle
path = kagglehub.dataset_download("divyansh1010/kdd-nids")

print("Path to dataset files:", path)

Path to dataset files: /kaggle/input/kdd-nids


In [ ]:
#Loading data
train = os.path.join(path, "KDD.csv")

df = pd.read_csv(train)

# Getting Info About Data

In [ ]:
#Printing the first 5 rows of dataset to get a basic idea
df.head()

duration protocol_type  service flag  src_bytes  dst_bytes  land  \
0         0           udp    other   SF        146          0     0   
1         0           tcp  private   S0          0          0     0   
2         0           tcp     http   SF        232       8153     0   
3         0           tcp     http   SF        199        420     0   
4         0           tcp  private  REJ          0          0     0   

   wrong_fragment  urgent  hot  ...  dst_host_same_srv_rate  \
0               0       0    0  ...                    0.00   
1               0       0    0  ...                    0.10   
2               0       0    0  ...                    1.00   
3               0       0    0  ...                    1.00   
4               0       0    0  ...                    0.07   

   dst_host_diff_srv_rate  dst_host_same_src_port_rate  \
0                    0.60                         0.88   
1                    0.05                         0.00   
2                    0.00                         0.03   
3                    0.00                         0.00   
4                    0.07                         0.00   

   dst_host_srv_diff_host_rate  dst_host_serror_rate  \
0                         0.00                  0.00   
1                         0.00                  1.00   
2                         0.04                  0.03   
3                         0.00                  0.00   
4                         0.00                  0.00   

   dst_host_srv_serror_rate  dst_host_rerror_rate  dst_host_srv_rerror_rate  \
0                      0.00                   0.0                      0.00   
1                      1.00                   0.0                      0.00   
2                      0.01                   0.0                      0.01   
3                      0.00                   0.0                      0.00   
4                      0.00                   1.0                      1.00   

     label  difficulty_score  
0   normal                15  
1  neptune                19  
2   normal                21  
3   normal                21  
4  neptune                21  

[5 rows x 43 columns]

In [ ]:
#Printing all the column names
df.columns

Index(['duration', 'protocol_type', 'service', 'flag', 'src_bytes',
       'dst_bytes', 'land', 'wrong_fragment', 'urgent', 'hot',
       'num_failed_logins', 'logged_in', 'num_compromised', 'root_shell',
       'su_attempted', 'num_root', 'num_file_creations', 'num_shells',
       'num_access_files', 'num_outbound_cmds', 'is_host_login',
       'is_guest_login', 'count', 'srv_count', 'serror_rate',
       'srv_serror_rate', 'rerror_rate', 'srv_rerror_rate', 'same_srv_rate',
       'diff_srv_rate', 'srv_diff_host_rate', 'dst_host_count',
       'dst_host_srv_count', 'dst_host_same_srv_rate',
       'dst_host_diff_srv_rate', 'dst_host_same_src_port_rate',
       'dst_host_srv_diff_host_rate', 'dst_host_serror_rate',
       'dst_host_srv_serror_rate', 'dst_host_rerror_rate',
       'dst_host_srv_rerror_rate', 'label', 'difficulty_score'],
      dtype='object')

In [ ]:
#Getting the Data type of each column and how many non null values they have
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148515 entries, 0 to 148514
Data columns (total 43 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   duration                     148515 non-null  int64  
 1   protocol_type                148515 non-null  object 
 2   service                      148515 non-null  object 
 3   flag                         148515 non-null  object 
 4   src_bytes                    148515 non-null  int64  
 5   dst_bytes                    148515 non-null  int64  
 6   land                         148515 non-null  int64  
 7   wrong_fragment               148515 non-null  int64  
 8   urgent                       148515 non-null  int64  
 9   hot                          148515 non-null  int64  
 10  num_failed_logins            148515 non-null  int64  
 11  logged_in                    148515 non-null  int64  
 12  num_compromised              148515 non-null  int64  
 13 

In [ ]:
#The shape of whole data
df.shape

(148515, 43)

# Handling Missing Values

In [ ]:
#Checking if there is any missing value in the data
print(df.isnull().sum())

duration                       0
protocol_type                  0
service                        0
flag                           0
src_bytes                      0
dst_bytes                      0
land                           0
wrong_fragment                 0
urgent                         0
hot                            0
num_failed_logins              0
logged_in                      0
num_compromised                0
root_shell                     0
su_attempted                   0
num_root                       0
num_file_creations             0
num_shells                     0
num_access_files               0
num_outbound_cmds              0
is_host_login                  0
is_guest_login                 0
count                          0
srv_count                      0
serror_rate                    0
srv_serror_rate                0
rerror_rate                    0
srv_rerror_rate                0
same_srv_rate                  0
diff_srv_rate                  0
srv_diff_h

In [ ]:
#Dropping in case any null value (in this case, no null value)
df = df.dropna()

In [ ]:
#Printing the dimensions again
df.shape

(148515, 43)

The shape didnt changed because there was no null values

# Handling Duplicate Values

In [ ]:
#Checking for any duplicated rows
df.duplicated().sum()

np.int64(610)

In [ ]:
#Dropping all the duplicate values
df.drop_duplicates(inplace=True)

In [ ]:
#Printing the dimensions again
df.shape

(147905, 43)

The rows got less by 610 because there were 610 duplicate values

# Data Description

In [ ]:
#Describing the data
df.describe().T

count          mean           std  min    25%  \
duration                     147905.0    277.924560  2.465704e+03  0.0   0.00   
src_bytes                    147905.0  40387.340874  5.420791e+06  0.0   0.00   
dst_bytes                    147905.0  17158.582955  3.711179e+06  0.0   0.00   
land                         147905.0      0.000196  1.400124e-02  0.0   0.00   
wrong_fragment               147905.0      0.020432  2.400420e-01  0.0   0.00   
urgent                       147905.0      0.000203  1.945721e-02  0.0   0.00   
hot                          147905.0      0.189960  2.017202e+00  0.0   0.00   
num_failed_logins            147905.0      0.004341  7.239701e-02  0.0   0.00   
logged_in                    147905.0      0.404307  4.907590e-01  0.0   0.00   
num_compromised              147905.0      0.256016  2.227731e+01  0.0   0.00   
root_shell                   147905.0      0.001514  3.888704e-02  0.0   0.00   
su_attempted                 147905.0      0.000980  4.247663e-02  0.0   0.00   
num_root                     147905.0      0.274859  2.273591e+01  0.0   0.00   
num_file_creations           147905.0      0.012123  5.189329e-01  0.0   0.00   
num_shells                   147905.0      0.000527  2.775774e-02  0.0   0.00   
num_access_files             147905.0      0.004030  9.545315e-02  0.0   0.00   
num_outbound_cmds            147905.0      0.000000  0.000000e+00  0.0   0.00   
is_host_login                147905.0      0.000081  9.007060e-03  0.0   0.00   
is_guest_login               147905.0      0.012359  1.104836e-01  0.0   0.00   
count                        147905.0     83.051966  1.165271e+02  0.0   2.00   
srv_count                    147905.0     27.938859  7.476439e+01  0.0   2.00   
serror_rate                  147905.0      0.257605  4.322833e-01  0.0   0.00   
srv_serror_rate              147905.0      0.256015  4.329466e-01  0.0   0.00   
rerror_rate                  147905.0      0.137650  3.390588e-01  0.0   0.00   
srv_rerror_rate              147905.0      0.138189  3.414580e-01  0.0   0.00   
same_srv_rate                147905.0      0.672711  4.366060e-01  0.0   0.10   
diff_srv_rate                147905.0      0.067839  1.947318e-01  0.0   0.00   
srv_diff_host_rate           147905.0      0.097681  2.591537e-01  0.0   0.00   
dst_host_count               147905.0    183.850215  9.850460e+01  0.0  86.00   
dst_host_srv_count           147905.0    119.389750  1.112051e+02  0.0  11.00   
dst_host_same_srv_rate       147905.0      0.533858  4.479554e-01  0.0   0.05   
dst_host_diff_srv_rate       147905.0      0.084216  1.941194e-01  0.0   0.00   
dst_host_same_src_port_rate  147905.0      0.144211  3.064946e-01  0.0   0.00   
dst_host_srv_diff_host_rate  147905.0      0.030006  1.063074e-01  0.0   0.00   
dst_host_serror_rate         147905.0      0.256800  4.288545e-01  0.0   0.00   
dst_host_srv_serror_rate     147905.0      0.251966  4.300847e-01  0.0   0.00   
dst_host_rerror_rate         147905.0      0.135922  3.223440e-01  0.0   0.00   
dst_host_srv_rerror_rate     147905.0      0.136092  3.349239e-01  0.0   0.00   
difficulty_score             147905.0     19.280653  2.742570e+00  0.0  18.00   

                                50%     75%           max  
duration                       0.00    0.00  5.771500e+04  
src_bytes                     44.00  278.00  1.379964e+09  
dst_bytes                      0.00  580.00  1.309937e+09  
land                           0.00    0.00  1.000000e+00  
wrong_fragment                 0.00    0.00  3.000000e+00  
urgent                         0.00    0.00  3.000000e+00  
hot                            0.00    0.00  1.010000e+02  
num_failed_logins              0.00    0.00  5.000000e+00  
logged_in                      0.00    1.00  1.000000e+00  
num_compromised                0.00    0.00  7.479000e+03  
root_shell                     0.00    0.00  1.000000e+00  
su_attempted                   0.00    0.00  2.000000e+00  
num_root   

In [ ]:
#Counting occurences of each unqiue column in the target column(label)
df['label'].value_counts()

label
normal             76966
neptune            45715
satan               4360
ipsweep             3643
smurf               3108
portsweep           3070
nmap                1566
back                1300
guess_passwd        1284
mscan                996
warezmaster          964
teardrop             901
warezclient          890
apache2              737
processtable         685
snmpguess            331
saint                319
mailbomb             293
pod                  221
snmpgetattack        178
httptunnel           133
buffer_overflow       50
multihop              25
rootkit               23
land                  22
named                 17
ps                    15
sendmail              14
xterm                 13
imap                  12
ftp_write             11
loadmodule            11
xlock                  9
phf                    6
perl                   5
xsnoop                 4
spy                    2
worm                   2
sqlattack              2
udpstorm               2
Name: count, dtype: int64

# Changing for binary labelling

In [ ]:
#Making a new column called 'binary_label' for binary calssifying label in 2 sets
df['binary_label'] = df['label'].apply(lambda x: 'normal' if x == 'normal' else 'attack')

# Exporting for further Use

In [ ]:
#Exporting this cleaned dataset for further use
df.to_pickle("cleaned_dataset.pkl")